### ISSUE TITLE
[NVIDIA] Performance - D3D11 titles vs D3D12 titles

### ISSUE THREAD

**Issue by shelterx** [HansKristian-Work/vkd3d-proton#1922]
It'll leave this here for cross reference, everything is in the forum post.
But in short, it feels like D3D12 suffer more then D3D11 titles with Nvidia GPUs, I understand there's technical differences between the two API's but it feels like something is off somewhere.

https://forums.developer.nvidia.com/t/please-sort-out-the-dx12-vkd3d-proton-performance-under-linux/284642/2

Thanks for all your efforts @doitsujin and @HansKristian-Work and everyone else involved.
If I can assist in any way, please let me know.

**Comment by Blisto91** [comment]
It is a known issue that Nvidia isn't working optimally together with vkd3d-proton compared to native d3d12. Things to improve it have and are still being explored.
https://github.com/HansKristian-Work/vkd3d-proton/pull/1902 Is some of the latest work in this category.

**Comment by Blisto91** [comment]
The above work can now be tested on Proton Bleeding Edge.
Right now this needs the Vulkan Dev driver to working (550.40.53)

**Comment by shelterx** [comment]
I compiled Proton Experimental Bleeding Edge.
Tried 3 games, saw no difference unfortunately. I expected to see some change at least.

**Comment by Blisto91** [comment]
It is expected to only have a noticable effect in some games. Dead Space is one known. Lies of P too I've gathered.

**Comment by shelterx** [comment]
Is there a way to check if nv_raw_access_chains is really being used?

**Comment by shelterx** [comment]
@doitsujin @HansKristian-Work 
Sorry to ping you but isn't this a bit weird, look at the GPU usage, from Control, this is with RT on. Obviously the card has a lot more to give...
![image](https://github.com/HansKristian-Work/vkd3d-proton/assets/8688458/a5755e4f-bbf5-46a1-bf7e-9503b29cf868)

Same location on Windows under D3D12 gives me ~90 FPS.

**Comment by shelterx** [comment]
Ok, so, if anyone can explain this? I noticed that turning Resizable Bar OFF (Above 4G decoding still ON) in BIOS actually makes the game use 99-100% of the GPU and bumps the FPS with ~15... so from 60, to 75. Still 15 FPS away from Windows but quite the bump nonetheless...

I also tested VKD3D_CONFIG=no_upload_hvv and Rebar ON, it did not make any difference at all.

**Comment by shelterx** [comment]
I did a test with Rebar off again but it didn't really help Control this time from what I could tell, then I disabled Above 4G decoding and boom, GPU usage was 100% and the FPS increased.

@HansKristian-Work Is there any way to bypass Above 4G decoding in vkd3d-proton? Or is it something managed by the driver?
There are reports that Alan Wake also benefits slightly with 4G decoding off.

**Comment by doitsujin** [comment]
You can try `VKD3D_CONFIG=no_upload_hvv`, which prevents us from putting certain memory allocations in VRAM on ReBAR-enabled systems.

**Comment by shelterx** [comment]
@doitsujin Yea, but no_upload_hvv does not help at all, disabling 4G Decoding entirely in BIOS seems to be the only fix. That's why I was wondering if there was some way to disable 4G decoding in vkd3d-proton somehow. At least then we'll have a togglable setting for it.

**Comment by HansKristian-Work** [comment]
No, we cannot override a BIOS setting from user space.

**Comment by shelterx** [comment]
Yeah, it's quite possible a driver issue too. I have no idea. I was hoping 4G decoding could be bypassed in the code,
Windows is not affected by this under DX12.

**Comment by shelterx** [comment]
A followup (again, I know)... I still feel like something is off here...
When starting Control vkd3d-proton log says this:
1f0:warn:d3d12_query_heap_create: Working around query heap bug.
01f0:warn:vkd3d_allocate_device_memory: Memory allocation failed, falling back to system memory.

That doesn't seem right? Why would it fall back to system memory when I have a 12GB card? I had an 8GB card some months ago and I thought it was on the edge for running Control in Linux, but it seems to be something else.

**Comment by shelterx** [comment]
Hmm, I had another look at this. Is this a purely a game issue? It does not seem to be able to allocate required memory if Rebar is enabled, is there any way around this or figure out what the heck is going on? I tried with no_upload_hvv, then the allocation warning disappears but the performance does not improve, GPU usage is still 73% and not 100%.
The performance only improves if I actually disable Rebar/4G decoding in bios.

**Comment by HansKristian-Work** [comment]
The "falling back" message can happen if ReBAR budget is exhausted. VKD3D_CONFIG=log_memory_budget gives a clearer picture.

**Comment by shelterx** [comment]
Here's the log.
[memory.log](https://github.com/user-attachments/files/16765413/memory.log)

System:
RTX 4070 - 12GB VRAM
Gigabyte Aorus Master Z390
Intel i9 9900KS
32GB RAM

**Comment by shelterx** [comment]
I tried vkd3d-proton under Windows with Control (RT on)...

While there is a maybe slight performance drop in Linux compared to Windows with vkd3d-proton (might be because of the proton layer itself), the drop can't really be blamed on the Nvidia Linux drivers, At least not when it's comes to this game,
The GPU utilization issue with Rebar is not present under Windows tho', I don't think the falling back to system memory is the issue, since it does that under windows too.

**Comment by shelterx** [comment]
I've tried to do some more reasearch...
Nvidias Vulkan performance doesn't seem to be bad using a native benchmark.
https://www.phoronix.com/review/breaking-limit-rt-linux

There's also this rather old post:
https://forums.developer.nvidia.com/t/linux-benchmarks-for-native-games/67238

I could be wrong but the performance degradation seems to be affecting vkd3d-proton only.

**Comment by shelterx** [comment]
Done some more testing on my end...
Native vulkan - Same or better than Windows (RDR2 and Breaking limit tested)
DXVK - ~the same as Windows running DX11, may vary slightly.
vkd3d-proton - Up to 30% performance loss vs DX12 depending on title, both in Linux and Windows.

The performance issues with vkd3d-proton isn't Linux driver specific, which I've always thought.

**Comment by shelterx** [comment]
Closing this, it got a bit hard to follow and I'm basically just having a monologue with myself. Besides more people started to see that Nvidia got issues with vkd3d-proton by now. There's a "new" thread on the Nvidia forums that gained some traction.

I'm curious where the bottleneck is tho', is it just "bad drivers" or does it go deeper than that and is somewhat a hardware limitation somewhere?
